common/lib/client/auth.js

```diff
@@ -534,6 +519,144 @@ var Auth = (function() {
 		}
 	};
 
+	Auth.prototype._saveBasicOptions = function(defaultTokenParams, authOptions) {
+		this.method = 'basic';
+		this.key = authOptions.key;
+		this.basicKey = toBase64(authOptions.key);
+		this.tokenParams = defaultTokenParams || {};
+		this.authOptions = authOptions || {};
+		delete this.authOptions.force;
+		if('clientId' in authOptions) {
+			this._userSetClientId(authOptions.clientId);
+		}
+	}
+
+	Auth.prototype._saveTokenOptions = function(tokenParams, authOptions) {
+		this.method = 'token';
+
+		this.tokenParams = tokenParams || this.tokenParams || {};
+
+		/* If an authOptions object is passed in that contains new auth info (ie
+		* isn't just {force: true} or something), it becomes the new default, with
+		* the exception of the force attribute (RSA10g), which is set anew on each
+		* call to authorise (defaulting to false) */
+		this.force = false;
+		if(authOptions) {
+			this.force = authOptions.force;
+
+			if(containsAuthInfo(authOptions)) {
+				this.authOptions = authOptions || {};
```

Comment:
> Assuming therefore the user did the following, I assume it would effectively nullify the clientId and the authUrl

In the implementation in this PR: no, to both. auth.authOptions is only overwritten if it `containsAuthInfo` (ie an authUrl, authCallback, token, tokenDetails, or key), to avoid this problem (see the line above the line this comment is on). `auth.clientId` is only ever touched if authOptions has a clientId key ([see here](https://github.com/ably/ably-js/pull/261/files#diff-cc36f4a032c8d16ad5390d989524c305R586)) so it would not get blanked out even for an authOptions that containsAuthInfo.

That said, I take the point, and don't mind changing them both to merge if @paddybyers agrees. (Though we'd have to make sure that's documented well, especially since with requestToken and createTokenRequest, a passed in tokenParams is explicitly not merged, but replaces any stored tokenParams)